### ISSUE TITLE
QSA persistent_topk makes greedy prefill non-deterministic on GB10; exact top-k fixes it

### ISSUE THREAD

**Issue by k3dani** [blazux/qwen3.8-Flash-DGX#3]
<h2><span>Summary</span></h2><p class="isSelectedEnd"><span>While evaluating this recipe on a single DGX Spark, I found that byte-identical sequential requests produced different prefill logits and different greedy outputs at </span><code dir="ltr"><span>temperature=0</span></code><span>.</span></p><p class="isSelectedEnd"><span>The issue has now been isolated to the Qwen3.8-Flash-Next QSA indexer’s </span><code dir="ltr"><span>persistent_topk</span></code><span> path on GB10 (</span><code dir="ltr"><span>sm_121a</span></code><span>). Replacing that selection with </span><code dir="ltr"><span>torch.topk</span></code><span> plus canonical tie ordering made the prefill bit-exact deterministic.</span></p><p class="isSelectedEnd"><span>This does </span><strong><span>not</span></strong><span> appear to be caused by this repository’s mmap-PLE patch: the patch does not modify the QSA code path. I am reporting it here because users of this single-Spark recipe are directly affected.</span></p><h2><span>Environment</span></h2><ul data-spread="false"><li><span>DGX Spark / GB10, </span><code dir="ltr"><span>sm_121a</span></code></li><li><span>Driver </span><code dir="ltr"><span>580.173.02</span></code></li><li><span>vLLM </span><code dir="ltr"><span>0.1.dev20073+g8e685d198</span></code></li><li><span>This repository at commit </span><code dir="ltr"><span>82ed48d</span></code></li><li><code dir="ltr"><span>RadixArk/Qwen3.8-Flash-Next-NVFP4</span></code></li><li><span>mmap PLE enabled</span></li><li><span>MTP=2</span></li><li><span>CUDA graphs: </span><code dir="ltr"><span>PIECEWISE</span></code></li><li><span>NVFP4 MoE backend: </span><code dir="ltr"><span>FLASHINFER_CUTLASS</span></code></li><li><span>One request in flight</span></li></ul><h2><span>Observed behavior</span></h2><p class="isSelectedEnd"><span>On a 50-item extraction suite, with every item sent three times using </span><code dir="ltr"><span>temperature=0</span></code><span> and </span><code dir="ltr"><span>top_p=1</span></code><span>:</span></p><ul data-spread="false"><li><strong><span>13/50</span></strong><span> items produced different greedy outputs</span></li><li><strong><span>5/50</span></strong><span> changed the extracted date or amount</span></li><li><span>on one item, majority voting selected the wrong result</span></li><li><span>divergence frequently started at the first reasoning token</span></li></ul><p class="isSelectedEnd"><span>Control configurations using llama.cpp, a 35B FP8 vLLM model, and a 122B NVFP4 vLLM model were stable.</span></p><h2><span>Isolation results</span></h2><ul data-spread="false"><li><span>MTP disabled: </span><strong><span>11/13</span></strong><span> problematic items remained unstable</span></li><li><code dir="ltr"><span>cudagraph_mode=NONE</span></code><span>: </span><strong><span>13/13</span></strong><span> remained unstable</span></li><li><span>canonical reordering of the existing </span><code dir="ltr"><span>persistent_topk</span></code><span> output: some probes remained unstable</span></li><li><span>exact </span><code dir="ltr"><span>torch.topk</span></code><span> plus canonical ordering: </span><strong><span>0/13</span></strong><span> unstable</span></li></ul><p class="isSelectedEnd"><span>A dedicated prefill probe sent six identical prompts ten times with:</span></p><pre dir="ltr"><code dir="ltr"><span>temperature=0
top_p=1
max_tokens=1
top_logprobs=20
concurrency=1</span></code></pre><p class="isSelectedEnd"><span>Results:</span></p>
QSA selection | Bit-identical top-20 vectors
--- | ---
Stock persistent_topk | 0/6 prompts
torch.topk + canonical ordering | 6/6 prompts

<p class="isSelectedEnd"><span>With the fix, the full 50-item suite completed with </span><strong><span>0/50 unstable items</span></strong><span>. Decode performance remained unchanged, while prefill throughput decreased from approximately 2,360 to 1,760 tokens/s.</span></p><p class="isSelectedEnd"><span>Reordering the kernel output alone was insufficient, indicating that the selected top-k </span><strong><span>set</span></strong><span> changes between launches, not only its output order. This is consistent with the existing </span><code dir="ltr"><span>persistent_topk</span></code><span> correctness report:</span></p><p class="isSelectedEnd"><a href="https://github.com/vllm-project/vllm/issues/51782" data-rich-text-autolink=""><span>https://github.com/vllm-project/vllm/issues/51782</span></a></p><h2><span>Reproduction, patch, and raw results</span></h2><p class="isSelectedEnd"><span>The complete reproducible experiment contains the probe, synthetic dataset, raw JSON results, exact top-k patch, Docker overlay, isolation scripts, and validation reports:</span></p><p class="isSelectedEnd"><a href="https://github.com/k3net/docai-evals/tree/master/experiments/2026-08-28-qwen38-flash-next-nvfp4-topk-nondeterminism-gb10" data-rich-text-autolink=""><span>https://github.com/k3net/docai-evals/tree/master/experiments/2026-08-28-qwen38-flash-next-nvfp4-topk-nondeterminism-gb10</span></a></p><h2><span>Detailed write-up</span></h2><p class="isSelectedEnd"><span>A readable explanation of the investigation, the kernel behavior, performance cost, and the separate MTP/reasoning-loop finding:</span></p><p class="isSelectedEnd"><a href="https://docai.hu/en/blog/qwen38-flash-next-nondeterministic-vllm-kernel"><span>Non-deterministic Qwen3.8-Flash-Next: the faulty vLLM kernel and its fix</span></a></p><h2><span>Suggested action</span></h2><p class="isSelectedEnd"><span>Until an upstream deterministic </span><code dir="ltr"><span>persistent_topk</span></code><span> implementation is available, it may be useful to:</span></p><ol data-spread="false" start="1"><li><span>document this as a known correctness limitation on GB10;</span></li><li><span>provide an optional deterministic exact-top-k mode for users who prioritize reproducibility;</span></li><li><span>run a simple first-token determinism probe as part of the smoke test.</span></li></ol><p><span>I am happy to test an upstream branch or alternative kernel on the DGX Spark.</span></p>

**Comment by blazux** [comment]
Confirmed on our GX10, and fixed in `main` as of 8347e7c — thanks for the precise report and the upstream issue (vllm#51782).

What we measured: with the stock `persistent_topk`, 3 identical greedy runs give 3 different completions on 2 of 4 prompts (2.6k–128k tokens), with first-token top-5 sets that don't overlap. Masking the never-written logits columns before the stock kernel does not help, so it is the kernel itself, as you found. GB10 is more exposed than other GPUs: `cooperative_topk` is disabled on sm_12x, so `persistent_topk` runs for decode too.

Now in the image: `VLLM_QSA_EXACT_TOPK=1` (`src/patch_qsa_exact_topk.py`), on by default in `scripts/serve.sh` — `torch.topk` over the visible columns, in place. Results: 4/4 prompts stable, first-token logprobs identical to the 4th decimal across runs, no change on our 17-scenario agentic tournament (44/51 before and after). Cost: decode unchanged, prefill −8% at 8k, −20–40% at 32k+. `EXACT_TOPK=0` keeps the stock kernel for anyone who prefers speed. `scripts/smoke-test.sh` now includes the determinism check you suggested, and there is a CPU unit test for the exact path. Credited in the README. Closing.

**Comment by jschmied** [comment]
A kernel-side fix for this is now a vLLM PR: vllm-project/vllm#55122 makes `persistent_topk` deterministic — index-ordered output, exact tie handling, no candidate buffers (so it also fixes vllm#51782, the dropped candidates).

Cost on GB10: 1.3–3× per call in the microbenchmark, ≈ +2 % TTFT at the model level by estimate, versus the −17…−40 % prefill that an exact `torch.topk` replacement costs (ours was +6 %). End-to-end decode/TTFT numbers follow on the PR.

If you want it on the preview tree without rebuilding vLLM: standalone build + env-gated wiring in https://github.com/jschmied/qwen38-flash-next-gb10/tree/main/patches/kernel-det (`build_det.py`, `tools/determinism/qsadet_patch.py`, `VLLM_QSA_DET_TOPK=1`).

For full reproducibility of this model on GB10 two more defects matter: the FlashInfer CUTLASS MoE fused finalize (vllm#54945, PR #54948) and the align-mode block-size PRs vllm#54076 / #53798.

**Comment by k3dani** [comment]
Thanks — this looks like the kernel-side fix we were hoping for. I’ll validate #55122 on the same single-GB10 RadixArk NVFP4 setup and compare it against our exact torch.topk baseline using the 6×10 first-token probe and the 50-item suite. I’ll also test it together with #54948 / use_fused_finalize=False, so the FlashInfer MoE path cannot mask the Top-K result. I’ll report determinism, full-token logprob hashes, and TTFT at 8K and 32K.

**Comment by jschmied** [comment]
If test is successful, feel free to comment at the PR to give it a bit more weight.

**Comment by k3dani** [comment]
Thanks — absolutely. I have completed the production-baseline phase first: 92 requests with token-level top-20 logprob hashing, including prefix caching, MTP=2 and up to 512 generated tokens. The warm serial runs were stable, but I found a separate partial-vs-full prefix-cache-state discrepancy that needs a controlled follow-up.

Next I’ll test #55122 against both stock persistent_topk and our exact torch.topk baseline, starting with max_tokens=1 to isolate the QSA prefill path. I’ll use three server starts and report determinism plus 8K/32K performance directly on the PR if the result is successful.

**Comment by jschmied** [comment]
added a paragraph to https://github.com/jschmied/qwen38-flash-next-gb10/blob/main/patches%2Fkernel-det%2FREADME.md

**Comment by blazux** [comment]
**Update: @jschmied's deterministic `persistent_topk` kernel is now the default in this repo (commit 4b723de).**

Tested on the GX10, hybrid checkpoint, MTP=2, prefix caching, YaRN 500k:

| | stock kernel | exact `torch.topk` (previous fix) | **deterministic kernel** |
|---|---|---|---|
| Deterministic (4 prompts × 3 runs, first-token logprobs) | no | yes (Δ 0.000) | **yes (Δ 0.000)** |
| Decode | 32.4 tok/s | 30.8 | **32.5** |
| Prefill 8k | ~1,650 (page-cache dependent) | 1,476 | **2,436** |
| Prefill 32k | 2,105 | 1,794 | **2,904** |
| Needle 92k | 45 s | 69 s | **48 s** |

His standalone test (177 cases incl. adversarial ties) passes 177/177 on the GB10, and the stock op fails to reproduce itself on the same inputs — so it is the kernel, as suspected.

What changed in the repo:
- Dockerfile patch 8 fetches the kernel sources from https://github.com/jschmied/qwen38-flash-next-gb10 at a pinned commit and compiles them with the image's `nvcc` at `docker build` (~12 s, no vLLM rebuild), wired into the QSA block selection via his `qsadet_patch.py` behind `VLLM_QSA_DET_TOPK=1`.
- `scripts/serve.sh`: `DET_TOPK=1` is the default. `EXACT_TOPK=1` (the `torch.topk` path) stays as a fallback and wins when set; `DET_TOPK=0 EXACT_TOPK=0` gives the stock kernel back.
- Details in the README ("Deterministic top-k") and docs/HOW-IT-WORKS.md.

Same determinism as the exact-top-k fix, whole prefill penalty gone. Thanks @jschmied for the kernel and @k3dani for the original diagnosis. Upstream is vllm-project/vllm#55122 — once it lands in the release the image is built from, patch 8 becomes redundant.

@jschmied : your repo has no LICENSE file yet 😕

**Comment by jschmied** [comment]
added Apache License 2.0 to sources

**Comment by jschmied** [comment]
you might also try https://github.com/vllm-project/vllm/pull/55180 to speed up larger prefill

**Comment by jschmied** [comment]
Two things measured on our GB10 this week that apply directly to this image, plus one thing not to spend a tournament on.

**1. The blockwise-FP8 GEMM's `M % 4` slow path (the preview image lacks vllm#52775).** On the image's vLLM the sm_12x blockwise-FP8 kernel takes a `swap_ab` route whenever the scheduled chunk's row count is not a multiple of 4 (`(M <= 64) || (M % 4 != 0)`), and that route is ~1.6× slower at prefill M. Upstream fixed it in C++ on 08-19, after the image was cut. Measured on the stock image, prefix cache off, no spec, same prompt ±1 token:

| chunk rows | batch 8192 | batch 4096 |
|---|---|---|
| 7,503 (mod 4 = 3) | 4.52 s | 3.05 s |
| 7,508 (mod 4 = 0) | **2.86 s** | **2.81 s** |
| 29,263 / 29,268 | 12.60 / **11.50 s** | 11.11 / 11.11 s |

Your hybrid mode routes the GDN and QSA side layers through this very kernel, so it is more exposed than the stock NVFP4 checkpoint. Drop-in, no rebuild: `tools/main/fp8_m4pad_patch.py` in our repo pads M to a multiple of 4 inside `apply_block_scaled_mm` (zero rows + unit scale rows, output sliced, opaque custom op so `torch.compile` does not specialise on M); env `VLLM_FP8_PAD_M4=0` turns it off. Same target file as the image's. Server-level validation on our preview venv (batch 8192, prefix cache off, no spec, one start each so far, a second is running): 8k **2.84 s** with the patch vs 5.03 s without (3 requests: 4.58/2.84/2.82 vs 5.03/6.64/3.51 — the unpatched kernel is also far noisier); 30k **11.28 s** vs 13.52 s. It also removes the mod-4 sensitivity of the batch size: batch 8192 becomes strictly better than 4096 again.

**2. The trailing prefix-cache block under MTP.** With any EAGLE-family drafter (MTP included) vLLM drops the last full block on every prefix-cache hit, so every agent turn re-prefills it. vllm#53388 (merged 09-01, main only) adds `disable_eagle_block_drop` to the speculative config. Measured on our main-build serve, MTP n=3, prefix caching on, 8-turn EOS-correct agent loop, three interleaved starts:

| | default | flag on |
|---|---|---|
| warm turns (3–8), mean of 18 | 2.05 s | **1.52 s (−26 %)** |
| cached tokens per warm turn | 4,800 | **6,400** |
| MTP acceptance | 53–56 % | 57–60 % |

Acceptance does not move. For this image it is a port of the four `vllm/` files of #53388 (`config/speculative.py`, `v1/core/kv_cache_utils.py`, `v1/core/sched/scheduler.py`, `v1/core/single_type_kv_cache_manager.py`); with `MTP=2 PREFIX_CACHE=1` as defaults it is the single largest per-turn win we know of for agent loops on this model.

**3. Correcting my pointer above: our M-chunking PR (vllm#55180) is not worth testing here.** It only fires when a weight exceeds the L2 (24 MiB on GB10); Flash-Next's largest FP8 side-layer weights are 25–30 MiB, so the bound is ~3 % and the server-level measurement was null (2.73 vs 2.71 s at 8k, 10.70 vs 10.64 s at 30k). It matters for dense models with 100+ MiB projections, not this one.

Both patches live at https://github.com/jschmied/qwen38-flash-next-gb10/tree/main/tools/main with the measurements in `notes/prefill-investigation.md` (findings 69/70, 89, 94).

**Comment by blazux** [comment]
Thanks @jschmied, this one was worth digging into — and it taught us something about our own defaults.

**Measured on the GX10** (hybrid, MTP=2, prefix caching on, YaRN 500k, `DET_TOPK=1`), same image with `VLLM_FP8_PAD_M4=0` as the control in the same session:

| | pad off | pad on |
|---|---|---|
| Decode | 31.6 tok/s | 32.7 |
| Prefill 8,001 tok | 3.33 s | 3.24 s |
| Prefill 32,001 tok | 11.63 s | 10.97 s |
| Prefill, salted prompts ×3 (no cache hits), 8k / 32k median | 3.51 / 13.88 s | 3.42 / 13.70 s |
| Needle 92k | 48.0 s | 47.2 s |
| Deterministic 4/4 | yes | yes |

So on our default config the effect is −3 to −6 % at best and within noise on the repeats — not the −40 % you saw. The kernel-level problem is real on our image though, and steeper than 1.6×. Micro-bench of the blockwise-fp8 cutlass GEMM (K=4096, N=8192): ×1.7 below 2,048 rows (0.63 → 1.09 ms at M=1,601), **×10–11 above** (0.87 → 9.6 ms at M=2,401, 1.59 → 16.2 ms at M=4,001). Your padding restores the aligned time in every case.

**Why it does not show for us: prefix caching.** You measured with the prefix cache off. With `--enable-prefix-caching` the scheduler's Mamba align mode (`_mamba_block_aligned_split`) clips every prefill chunk to the 1,600-token block boundary, so all large chunks reach the GEMM with M % 4 == 0; only the last chunk of a prompt has an arbitrary row count. We even built prompts to leave a misaligned last chunk above the 2,048-row cliff (8,800 / 8,801 / 8,803 tokens): 3.60 / 3.69 / 3.65 s, no difference. With the prefix cache off the chunks are whatever the batch size leaves, and your numbers are what people should expect.

**What we did:** shipped it as patch 9 in the image, opt-in (`PAD_M4=1`, hybrid mode with `PREFIX_CACHE=0` is the use case), documented in the README with the numbers above and credit to you (commit 89b852f). Our own box stays as it is. One note for anyone copying the patch by hand: it defaults to ON on sm_12x when `VLLM_FP8_PAD_M4` is unset, so `scripts/serve.sh` always passes the variable explicitly.

On #53388 (`disable_eagle_block_drop`): agreed that is the interesting one for agent loops, and your finding 98 (pad the static system prompt so the shared prefix ends on a 1,600-token boundary) is a nice zero-patch complement to it. The flag does not exist in the preview image, so it is a port of the four core files; we will look at it carefully given the block-size history on this path, with state checksums and a full bench rather than a TTFT bench. Noted that #55180 does not apply here.
